contracts/MatchPayouts.sol

```diff
@@ -0,0 +1,159 @@
+
+  /**
+   * @param _owner Address of contract owner
+   * @param _funder Address of funder
+   * @param _dai DAI address
+   */
+  constructor(address _owner, address _funder, IERC20 _dai) {
+    owner = _owner;
+    funder = _funder;
+    dai = _dai;
+  }
+
+  /// @dev Requires caller to be the owner
+  modifier onlyOwner() {
+    require(msg.sender == owner, "MatchPayouts: caller is not the owner");
+    _;
+  }
+
+  /// @dev Only allows method to be called once the payouts mapping has been finalized
+  modifier onlyWhenFinalized() {
+    require(finalized, "MatchPayouts: Payouts not finalized");
+    _;
+  }
+
+  // ======================================= PRIMARY METHODS =======================================
+  // Functions are laid out in the order they will be called over the lifetime of the contract
+
+  /**
+   * @notice Set's the mapping of addresses to their match amount
+   * @dev This will need to be called multiple times to prevent exceeding the block gas limit, based
+   * on the number of grants
+   * @param _payouts Array of `Payout`s to set
+   */
+  function setPayouts(Payout[] calldata _payouts) external onlyOwner {
+    require(!finalized, "MatchPayouts: Payouts already finalized");
+    // Set each payout amount
+    for (uint256 i = 0; i < _payouts.length; i += 1) {
+      payouts[_payouts[i].recipient] = _payouts[i].amount;
+      emit PayoutAdded(_payouts[i].recipient, _payouts[i].amount);
+    }
+  }
+
+  /**
+   * @notice Called by the owner to signal that the payout mapping is finalized
+   * @dev Once called, this cannot be reversed!
+   * @dev We use an explicit method here instead of doing this as part of the `setPayouts()` method
+   * to reduce the chance of accidentally setting this flag
+   */
+  function finalize() external onlyOwner {
+    finalized = true;
+    emit Finalized();
+  }
+
+  /**
+   * @notice Enables funder to withdraw all funds
+   * @dev Escape hatch, intended to be used if the payout mapping is finalized incorrectly. In this
+   * case a new MatchPayouts contract can be deployed and that one will be used instead
+   */
+  function withdrawFunding() external {
+    require(msg.sender == funder, "MatchPayouts: caller is not the funder");
+    dai.safeTransfer(funder, dai.balanceOf(address(this)));
+    emit FundingWithdrawn();
+  }
+
+  /**
+   * @notice Called by the owner to enable withdrawals of match payouts
+   * @dev Once called, this cannot be reversed!
+   */
+  function enablePayouts() external onlyOwner onlyWhenFinalized {
+    funded = true;
+    emit Funded();
+  }
+
+  /**
+   * @notice Withdraws funds owed to `_recipient`
+   * @param _recipient Address to withdraw for
+   */
+  function claimMatchPayout(address _recipient) external onlyWhenFinalized {
+    require(funded, "MatchPayouts: Not yet funded");
+    dai.safeTransfer(_recipient, payouts[_recipient]);
+    payouts[_recipient] = 0; // safe to put effect after interaction since we trust the Dai contract
```

Comment:
I figured since we trust Dai, may as well do it this way to reduce gas cost slightly. Just checked and looks like the change in gas cost is negligible, so changed this to use CEI in ad0da95